### DISCUSSION TITLE
"Failed to decrypt field" after upgrading to Renovate >= 37.317.0

### DISCUSSION THREAD

**Discussion by 9numbernine9** [renovatebot/renovate#28934] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Bitbucket Cloud; Renovate >= 37.317.0

### Please tell us more about your question or problem

Hello!

Since upgrading our self-hosted Renovate to versions >= `37.317.0`, we've started noticing a difference in behaviour with respect to encryption/decryption with GPG keys. (We've tested with `37.342.2`, `37.340.9`, and `37.328.1`). Previously the GPG keys that we were using worked perfectly fine, but after the `37.317.0` release they stopped working correctly; in our logs we would see `"Failed to decrypt field password. Please re-encrypt and try again."` errors, despite the keys working correctly previously.

Poking through the [release notes for `37.317.0`](https://github.com/renovatebot/renovate/releases/tag/37.317.0), we noticed that the PGP library was replaced with Renovate's fork of `kbpgp`. This release also added the `RENOVATE_X_USE_OPENPGP` flag, which we enabled for testing purposes and our encryption/decryption worked correctly once again. Knowing that this is an _experimental_ flag, however, this doesn't seem like a permanent solution.

What we've observed is that `kbpgp` seems to treat keys that are generated with a version of GnuPG that's _too new_ as being invalid. We tested this by generating a brand new RSA 4096-bit keypair using GnuPG `2.4.5`, re-encrypting all of our secrets with this new key, disabling the `RENOVATE_X_USE_OPENPGP` flag, and trying again - same decryption error as before. Then, we went through the same process but we used the `gpg` executable inside of the `renovatebot/renovate:37.342.2` image instead - which uses GnuPG `2.2.17` - re-encrypting all of our secrets _again_ with these new keys, and everything worked fine with `kbpgp` decryption.

With all of that said, my questions are:
- Has anyone else run into a similar issue? (Feel free to comment on this discussion.)
- Is this a known issue? (I couldn't find anything relevant on the issue tracker.)

My concern is that anyone who's using a newer GnuPG might run into this issue, and with the recent release of Ubuntu 24.04 - which defaults to GnuPG `2.4.x` - the chances of someone running into this situation are higher. 

### Logs (if relevant)

<details><summary>Logs</summary>

```

INFO: Repository has invalid config (repository=myorg/myrepo)
       "error": {
         "validationError": "Failed to decrypt field password. Please re-encrypt and try again.",
         "message": "config-validation",
         "stack": "Error: config-validation\n    at decryptConfig (/usr/local/renovate/lib/config/decrypt.ts:156:27)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at decryptConfig (/usr/local/renovate/lib/config/decrypt.ts:202:13)\n    at mergeRenovateConfig (/usr/local/renovate/lib/workers/repository/init/merge.ts:212:24)\n    at getRepoConfig (/usr/local/renovate/lib/workers/repository/init/config.ts:14:12)\n    at initRepo (/usr/local/renovate/lib/workers/repository/init/index.ts:55:12)\n    at Object.renovateRepository (/usr/local/renovate/lib/workers/repository/index.ts:62:14)\n    at attributes.repository (/usr/local/renovate/lib/workers/global/index.ts:217:11)\n    at start (/usr/local/renovate/lib/workers/global/index.ts:202:7)\n    at /usr/local/renovate/lib/renovate.ts:18:22"

```

</details>

**Comment by viceice** [comment]
- does it work with an old key and gnupg v2.4?
- doest it work with the new key and gnugp v2.2?

**Reply by 9numbernine9** [reply]
@viceice I've made a table of the combinations that I could reasonably try:

<table>
<tr>
<th rowspan="2">Keypair Created With</th>
<th rowspan="2">Secret Encrypted With</th>
<th colspan="2">Renovate</th>
</tr>
<tr>
<td><tt>RENOVATE_X_USE_OPENPGP</tt></td>
<td><tt>kbpgp</tt></td>
</tr>
<tr>
<td>GnuPG 2.2.17</td>
<td>GnuPG 2.2.17</td>
<td align="center" valign="center" style="text-align: center">✅</td>
<td align="center" valign="center" style="text-align: center">✅</td>
</tr>
<tr>
<td>GnuPG 2.2.17</td>
<td>GnuPG 2.4.5</td>
<td align="center" valign="center" style="text-align: center">✅</td>
<td align="center" valign="center" style="text-align: center">✅</td>
</tr>
<tr>
<td>GnuPG 2.4.5</td>
<td>GnuPG 2.2.17</td>
<td align="center" valign="center" style="text-align: center">✅</td>
<td align="center" valign="center" style="text-align: center">❌</td>
</tr>
<tr>
<td>GnuPG 2.4.5</td>
<td>GnuPG 2.4.5</td>
<td align="center" valign="center" style="text-align: center">✅</td>
<td align="center" valign="center" style="text-align: center">❌</td>
</tr>
</table>

When using `kbpgp` and with `LOG_LEVEL=debug`, the one interesting message I see in the logs is `"Unknown signature subpacket: 34"`, which doesn't appear when `RENOVATE_X_USE_OPENPGP` is enabled. It seems (to me) that there's something about keys generated with GnuPG >= `2.4` that isn't compatible with the `kbpgp` implementation. (I am not enough of a GnuPG/OpenPGP expert to know what this might be!)

**Reply by viceice** [reply]
ok, I'll check if we can fix our kbpgp fork.

@rarkins should we deprecate pgp in favor of RSA if not fixable?

**Reply by viceice** [reply]
https://github.com/keybase/keybase-issues/issues/4025

**Reply by viceice** [reply]
@9numbernine9 are you using RSA keys or ECC?

**Reply by rarkins** [reply]
Maybe can you create a new/dummy key pair which reproduces the problem and which you can share with us?

**Reply by viceice** [reply]
this workaround works for me:
```
% gpg --edit-key mykey

gpg> showpref
[ultimate] (1). mykey
     Cipher: AES256, AES192, AES, 3DES
     AEAD: OCB, EAX
     Digest: SHA512, SHA384, SHA256, SHA224, SHA1
     Compression: ZLIB, BZIP2, ZIP, Uncompressed
     Features: MDC, AEAD, Keyserver no-modify

gpg> setpref AES256 AES192 AES 3DES SHA512 SHA384 SHA256 SHA224 SHA1 ZLIB BZIP2 ZIP
Set preference list to:
     Cipher: AES256, AES192, AES, 3DES
     AEAD: 
     Digest: SHA512, SHA384, SHA256, SHA224, SHA1
     Compression: ZLIB, BZIP2, ZIP, Uncompressed
     Features: MDC, Keyserver no-modify
Really update the preferences? (y/N) y
```

You can only use rsa, no ecc keys.

**Reply by viceice** [reply]
ecc works too, sorry. was a typo on my tests

**Reply by viceice** [reply]
- https://github.com/renovatebot/kbpgp/pull/22

**Reply by 9numbernine9** [reply]
@viceice @rarkins Thank you for digging into this issue further! Apologies for not responding sooner. I am indeed using RSA 4096-bit keys.

@viceice I believe your workaround is indeed the correct course of action. Removing `AEAD` support from my key, then re-encrypting my secret with the key and giving the non-AEAD version of my private key to Renovate made everything work just fine, regardless if I used GnuPG `2.2` or `2.4` for key generation. The only practical difference between the two is that keys generated with `2.4` appear to choose AEAD by default; now that I _know_ what the problem is, I can see this has caused issues for others too. (Examples [here](https://github.com/open-keychain/open-keychain/issues/2886) and [here](https://security.stackexchange.com/questions/275883/should-one-really-disable-aead-for-recent-gnupg-created-pgp-keys/275891#275891) ; Arch Wiki [also recommends disabling it](https://wiki.archlinux.org/title/GnuPG#Disable_unsupported_AEAD_mechanism) FWIW).

Ideally I would love to see AEAD support added to `kbpgp` but I realize this might be non-trivial. Perhaps for now the best solution would be to update the self-hosted documentation to explicitly mention the _gotcha_ with AEAD?

**Reply by viceice** [reply]
Sadly `kbpgp` is not actively developed, so chances are bad to get it working. 

I've not enough knowledge to do the implementation and the Coffeescript language is making it even more worse.

I've also not enough time to convert it to regular JavaScript or Typescript for easier development. sorry 😔

**Comment by lildadou** [comment]
I would like to take advantage of this thread to mention a problem I encountered: if your ciphertext is less than 500 characters long, you will encounter an error during decryption. This can happen when your key is 2048 bits and your secret is short.

This problem is caused by this condition :
https://github.com/renovatebot/renovate/blob/01d4d0db8af4037d619e04c7ae21848acf9e6682/lib/config/decrypt/kbpgp.ts#L9-L12

**Comment by viceice** [comment, marked as answer]
It should now be possible to use AEAD with renovate v42 when setting `RENOVATE_X_PGP_RUNTIME=wasm-java` which will become the default on renovate v43

- #40678
- https://docs.renovatebot.com/self-hosted-experimental/#renovate_x_pgp_runtime
